core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+
+        // The goal is to traverse the entire tree
+        // formed by all path proofs and only collect the siblings
+        // that cannot be reconstructed in the future
+        //
+        // The traversal does not occur on the siblings themselves,
+        // but on the ordered set of key_paths within PathProofs.
+        //
+        // For example, take two key_paths starting from the first bit.
+        // If two keys share the same bit at index i, they will have the same sibling at index i.
+        // If the bit at index i differs, their paths diverge, and no sibling is needed at that index
+        // because it can be reconstructed from siblings collected later in the other key.
+        // When a differing bit is found, each key_path needs separate scanning.
+        //
+        // When there is a single key_path, all siblings are necessary.
+        //
+        // When there are more than two key_paths, the same logic is applied,
+        // but if two key_paths have different bits at an index, the key_paths are split
+        // based on the value at that index.
+        //
+        // If all key_paths share a bit at an index, that sibling is required;
+        // if at least two key_paths differ, no sibling is needed, but the key_paths must be divided
+        // based on having bit 0 or 1 at that index.
+        //
+        // Iterate this algorithm on the bisection to determine the minimum necessary siblings.
+        //
+        // `external_siblings` whill follow this structure for each bicection:
+        // |commons siblings| ext siblings in the right bisection | ext siblings in the left bisection |
+
+        // it represent a slice of the vector of key_paths with the notion of
+        // path_index, at which bit in the key_path that slice is pointing at
+        struct TraverseInfo {
+            path_index: usize,
+            path_slice_lower: usize,
+            path_slice_upper: usize,
+            ext_sibling_index: usize,
+        }
+        let mut stack_traverse_info: Vec<TraverseInfo> = vec![];
+
+        // External siblings encountered during a subtree traversal
+        let mut ext_sibling_in_sub_tree: Vec<Node> = vec![];
+
+        // initially we're looking at all the key_paths
+        let mut curr_traverse_info = TraverseInfo {
+            path_index: 0,
+            path_slice_lower: 0,
+            path_slice_upper: path_proofs.len() - 1,
+            ext_sibling_index: 0,
+        };
+
+        loop {
+            // Based on how we bisect the key_path slice, we can see that each TraverseInfo
+            // will represent a slice where all key_paths will have same bits before
+            // TraverseInfo.path_index.
+
+            // If there is no longer a common prefix, there is only one key in the TraverseInfo,
+            // and then all remaining siblings are required for the multiproof.
+            if curr_traverse_info.path_slice_lower == curr_traverse_info.path_slice_upper {
```

Comment:
this whole block could be extracted to a function like `fn prove_unique_path_remainder`